init-container/init.c

```diff
@@ -0,0 +1,361 @@
+        return -1;
+    }
+
+    char* buf = malloc(size + (is_cstring ? 1 : 0));
+    if (!buf) {
+        return -1;
+    }
+
+    if (readn(fd, buf, size) < 0) {
+        int tmp_errno = errno;
+        free(buf);
+        errno = tmp_errno;
+        return -1;
+    }
+
+    if (is_cstring) {
+        buf[size] = '\0';
+    }
+
+    *buf_ptr = buf;
+    *size_ptr = size;
+
+    return 0;
+}
+
+static void free_strings_array(char** array) {
+    if (array) {
+        for (size_t i = 0; array[i]; ++i) {
+            free(array[i]);
+        }
+        free(array);
+    }
+}
+
+static int recv_strings_array(int fd, char*** array_ptr) {
+    uint64_t size = 0;
+    int ret = -1;
+
+    if (recv_u64(fd, &size) < 0) {
+        return -1;
+    }
+
+    char** array = calloc(size + 1, sizeof(*array));
+    if (!array) {
+        return -1;
+    }
+
+    for (uint64_t i = 0; i < size; ++i) {
+        uint64_t tmp;
+        if (recv_bytes(fd, &array[i], &tmp, /*is_cstring=*/true) < 0) {
+            goto out;
+        }
+    }
+
+    *array_ptr = array;
+    array = NULL;
+    ret = 0;
+
+out:
+    if (array) {
+        int tmp_errno = errno;
+        free_strings_array(array);
+        errno = tmp_errno;
+    }
+    return ret;
+}
+
+static int writen(int fd, void* buf, size_t size) {
+    while (size) {
+        ssize_t ret = write(fd, buf, size);
+        if (ret == 0) {
+            puts("Waiting for host connection ...");
+            sleep(1);
+            continue;
+        }
+        if (ret < 0) {
+            /* `errno` should be set. */
+            return -1;
+        }
+        buf = (char*)buf + ret;
+        size -= ret;
+    }
+    return 0;
+}
+
+static void send_response(msg_id_t msg_id, uint32_t ret_val) {
+    struct msg_hdr resp = {
+        .msg_id = msg_id,
+        .type = !ret_val ? RESP_OK : RESP_ERR,
+    };
+
+    CHECK(writen(g_cmds_fd, &resp, sizeof(resp)));
+    if (ret_val) {
+        CHECK(writen(g_cmds_fd, &ret_val, sizeof(ret_val)));
+    }
+}
+
+static void print_strings_array(char** strings) {
+    for (unsigned i = 0; *strings; strings++, i++) {
+        printf("%u: %s\n", i, *strings);
+    }
+}
+
+static uint32_t spawn_new_process(uid_t uid, char* bin, char** argv,
+                                  char** envp) {
+    if (!envp) {
+        envp = environ;
+    }
+    printf("uid: %u\n", uid);
+    printf("Binary: %s\n", bin);
+    puts("ARGV:");
+    print_strings_array(argv);
+    puts("ENV:");
+    print_strings_array(envp);
+
+    return 0;
+}
+
+static void handle_run_process(msg_id_t msg_id) {
+    bool done = false;
+    char* bin = NULL;
+    char** argv = NULL;
+    char** envp = NULL;
+    uint32_t uid = DEFAULT_UID;
+
+    while (!done) {
+        uint8_t subtype = 0;
+        CHECK(readn(g_cmds_fd, &subtype, sizeof(subtype)));
+        switch (subtype) {
+            case SUB_MSG_END:
+                done = true;
+                break;
+            case SUB_MSG_RUN_PROCESS_BIN: ; // got to love C
```

Comment:
explain your affection ;)

Reply:
maybe you should do it like this:
```c
case SUB_MSG_RUN_PROCESS_BIN:
    {
        uint64_t bin_len = 0;
        CHECK(recv_bytes(g_cmds_fd, &bin, &bin_len, /*is_cstring=*/true));
    }
    break;
case [...]
```

Reply:
I do not like the superfluous nested scope and prefer simple empty statement (i.e. as it is done now).

Reply:
the problem is that the variable declaration applies to the whole `switch` statement. I like to use additional scope in switches, to keep variables local to a single case part.

Reply:
True, just moved the variable out of switch.